pkg/kmm/preflightvalidation.go

```
@@ -0,0 +1,327 @@
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the preflightvalidation is empty")
+
+		return builder, fmt.Errorf("%s", "preflightvalidation 'nsname' cannot be empty")
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("preflightvalidation object %s doesn't exist in namespace %s",
+			name, nsname)
+	}
+
+	builder.Definition = builder.Object
+
+	return builder, nil
+}
+
+// Create builds preflightvalidation in the cluster and stores object in struct.
+func (builder *PreflightValidationBuilder) Create() (*PreflightValidationBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating preflightvalidation %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	if !builder.Exists() {
+		err = builder.apiClient.Create(context.TODO(), builder.Definition)
+		if err == nil {
+			builder.Object = builder.Definition
+		}
+	}
+
+	return builder, err
+}
+
+// Update modifies an existing preflightvalidation on the cluster.
+func (builder *PreflightValidationBuilder) Update() (*PreflightValidationBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Updating preflightvalidation %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	err := builder.apiClient.Update(context.TODO(), builder.Definition)
+
+	if err == nil {
+		builder.Object = builder.Definition
+	}
+
+	return builder, err
+}
+
+// Exists checks if the defined preflightvalidation has already been created.
+func (builder *PreflightValidationBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if preflightvalidation %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
```

this may cause issues to have exists return true for other errors ("the Exists issue") but since we're going to fix that within the 4.20 timeframe anyway not a huge deal for this PR

I'd recommend logging the error in `Get` so it's not a silent failure but if it's not something you run into it's not a huge deal

---

Ok. I'll leave it like that and sort it out in all the `Exists` methods inside `kmm` pkg in a future PR